features/cellular/framework/mux/cellular_mux.cpp

```diff
@@ -0,0 +1,1237 @@
+            if (status == MUX_ESTABLISH_SUCCESS) {
+                _state.is_mux_open = 1u;
+            }
+
+            break;
+        default:
+            /* Code that should never be reached. */
+            MBED_ASSERT(false);
+            break;
+    };
+
+    _state.is_mux_open_running = 0;
+
+    return MUX_STATUS_SUCCESS;
+}
+
+
+void Mux::user_information_construct(uint8_t dlci_id, const void* buffer, size_t size)
+{
+    frame_hdr_t *frame_hdr =
+        reinterpret_cast<frame_hdr_t *>(&(Mux::_tx_context.buffer[FRAME_FLAG_SEQUENCE_FIELD_INDEX]));
+
+    frame_hdr->address  = EA_BIT | CR_BIT | (dlci_id << 2);
+    frame_hdr->control  = FRAME_TYPE_UIH;
+    frame_hdr->length   = (LENGTH_INDICATOR_OCTET | (size << 1));
+
+    memmove(&(frame_hdr->information[0]), buffer, size);
+
+    uint8_t* fcs_pos = (&(frame_hdr->information[0]) + size);
+    *fcs_pos         = fcs_calculate(&(Mux::_tx_context.buffer[FRAME_ADDRESS_FIELD_INDEX]), FCS_INPUT_LEN);
+    *(++fcs_pos)     = FLAG_SEQUENCE_OCTET;
+
+    _tx_context.bytes_remaining = UIH_FRAME_MIN_LEN + size;
+}
+
+
+void Mux::tx_noretransmit_entry_run()
+{
+    write_do();
+}
+
+
+ssize_t Mux::user_data_tx(uint8_t dlci_id, const void* buffer, size_t size)
+{
+    MBED_ASSERT(size <=
+                (MBED_CONF_MUX_BUFFER_SIZE - (FRAME_START_READ_LEN + FRAME_HEADER_READ_LEN + FRAME_TRAILER_LEN)));
+    MBED_ASSERT(buffer != NULL);
+
+    if (size == 0) {
+        return 0;
+    }
+
+    _mutex.lock();
+
+    ssize_t write_ret;
+    switch (_tx_context.tx_state) {
+        case TX_IDLE:
+            if (!_state.is_tx_callback_context) {
+                /* Proper state to start TX cycle within current context. */
+
+                user_information_construct(dlci_id, buffer, size);
+                tx_state_change(TX_NORETRANSMIT, tx_noretransmit_entry_run, tx_idle_exit_run);
+
+                write_ret = size;
+            } else {
+                /* Current context is TX callback context. */
+
+                if (!_state.is_user_tx_pending) {
+                    /* Signal callback context to start TX cycle and construct the frame. */
+
+                    /* @note: If user is starting a TX to a DLCI, which is after the current DLCI TX callback within the
+                     * stored sequence this will result to dispatching 1 unnecessary TX callback, if this is a issue one
+                     * should clear the TX callback pending bit marker for this DLCI in this place. */
+
+                    _state.is_user_tx_pending = 1u;
+                    user_information_construct(dlci_id, buffer, size);
+
+                    write_ret = size;
+                } else {
+                    /* TX cycle allready scheduled, set TX callback pending and inform caller by return value that no
+                       action was taken. */
+
+                    tx_callback_pending_bit_set(dlci_id);
+                    write_ret = 0;
```

Comment:
-EAGAIN is the correct return value for "I'm not sending anything because buffers full". Write never normally returns 0 unless asked to send 0.

Reply:
Managed by IOTCELL-567